File: let-nodejs/input/socrates.js
// inference.js
// A tiny forward-chaining inference engine for the predicate type/2
// Run with:  node inference.js

// ---------------------------------------------------------------------------
// knowledge base (set of [subject, category] tuples)
// ---------------------------------------------------------------------------
const facts = new Set();

/**
 * Serialises a tuple so it can live in a Set.
 */
function key(subject, category) {
  return `${subject}::${category}`;
}

/**
 * Assert an explicit fact  type(S, C).
 */
function assertType(subject, category) {
  facts.add(key(subject, category));
}

/**
 * Apply the rule  type(X,'Man') ⇒ type(X,'Mortal')  until no new facts arise.
 */
function infer() {
  let changed = true;
  while (changed) {
    changed = false;

    for (const k of Array.from(facts)) {
      const [subject, category] = k.split("::");
      if (category === "Man") {
        const mortalKey = key(subject, "Mortal");
        if (!facts.has(mortalKey)) {
          facts.add(mortalKey);
          changed = true;
        }
      }
    }
  }
}

/**
 * Equivalent to the Prolog query  ?- type(_, _).
 * Returns an array of [Subject, Category] pairs, sorted for stable output.
 */
function queryType() {
  infer();
  return Array.from(facts)
    .map(k => k.split("::"))
    .sort(([a1, b1], [a2, b2]) =>
      a1 === a2 ? b1.localeCompare(b2) : a1.localeCompare(a2)
    );
}

// ---------------------------------------------------------------------------
// Seed the knowledge base and run the “?- type(_, _).” query
// ---------------------------------------------------------------------------
assertType("Socrates", "Man");

if (require.main === module) {
  for (const [subject, category] of queryType()) {
    console.log(`type('${subject}', '${category}')`);
  }
}

File: let/input/socrates.py
# --- simple inference engine -----------------------------------------------

# storage for asserted facts
_facts: set[tuple[str, str]] = set()

def assert_type(subject: str, category: str) -> None:
    """Add a fact type(S, C) to the knowledge base."""
    _facts.add((subject, category))

def _infer() -> None:
    """Apply the rule   type(X,'Man') → type(X,'Mortal')   until saturation."""
    changed = True
    while changed:                      # keep looping until no new facts
        changed = False
        new_facts = {
            (s, "Mortal")
            for (s, c) in _facts
            if c == "Man" and (s, "Mortal") not in _facts
        }
        if new_facts:                   # at least one new fact was derived
            _facts.update(new_facts)
            changed = True

def query_type() -> list[tuple[str, str]]:
    """Return every known (explicit + inferred) type/2 fact."""
    _infer()
    return sorted(_facts)
# ---------------------------------------------------------------------------

# seed the knowledge base with the fact from your program
assert_type("Socrates", "Man")

if __name__ == "__main__":
    # equivalent to the Prolog query: ?- type(_, _).
    for fact in query_type():
        print(f"type{fact}")
